node/network/availability-distribution/src/lib.rs

```diff
@@ -166,13 +166,30 @@ struct PerCandidate {
 
 	/// The set of relay chain blocks this appears to be live in.
 	live_in: HashSet<Hash>,
+
+	/// A Jaeger span relating to this candidate.
+	span: jaeger::JaegerSpan,
 }
 
 impl PerCandidate {
 	/// Returns `true` iff the given `validator_index` is required by the given `peer`.
-	fn message_required_by_peer(&self, peer: &PeerId, validator_index: &ValidatorIndex) -> bool {
-		self.received_messages.get(peer).map(|v| !v.contains(validator_index)).unwrap_or(true)
-			&& self.sent_messages.get(peer).map(|v| !v.contains(validator_index)).unwrap_or(true)
+	fn message_required_by_peer(&self, peer: &PeerId, validator_index: ValidatorIndex) -> bool {
+		self.received_messages.get(peer).map(|v| !v.contains(&validator_index)).unwrap_or(true)
+			&& self.sent_messages.get(peer).map(|v| !v.contains(&validator_index)).unwrap_or(true)
+	}
+
+	/// Add a chunk to the message vault. Overwrites anything that was already present.
+	fn add_message(&mut self, chunk_index: u32, message: AvailabilityGossipMessage) {
+		let _ = self.message_vault.insert(chunk_index, message);
+	}
+
+	/// Clean up the span if we've got our own chunk.
+	fn drop_span_after_own_availability(&mut self) {
+		if let Some(validator_index) = self.validator_index {
+			if self.message_vault.contains_key(&validator_index) {
+				self.span = jaeger::JaegerSpan::Disabled;
```

Comment:
A bit of a quirk, I'll add a PR to handle this as part of the `JaegerSpan` API.

Reply:
After a few attempts, it seems this is the most idiomatic for now.

Reply:
You could maybe write an API like

```rust
impl JaegerSpan {
  fn drop_if(&mut self, condition: impl FnOnce() -> bool) { 
    if condition() {
      *self = JaegerSpan::Disabled;
    }
}
```

Which would mean that we could rewrite `drop_span_after_own_availability` as

```rust
self.span.drop_if(|| match self.validator_index {
  None => false,
  Some(validator_index) => self.message_vault.contains_key(&validator_index),
});
```

It's not a huge improvement over what currently exists, but it's something.

Reply:
@coriolinus Your snippet showing use of `drop_if` will not compile due to a borrow conflict on `self`. I think a simple `disable()` function would be sufficient to avoid passing things around in closures.